Batteries/Data/Fin/Coding.lean

```diff
@@ -0,0 +1,548 @@
+  simp only [decodeChar, encodeChar]
+  split
+  · simp [*]
+  · have : ¬ x.val + (0xE000 - 0xD800) < 0xD800 := by grind
+    simp [*]
+
+@[simp] theorem decodeChar_encodeChar (x) : decodeChar (encodeChar x) = x := by
+  ext; simp only [decodeChar, encodeChar]
+  split
+  · simp only [Char.ofNatAux, Char.toNat]; rfl
+  · have : ¬ x.toNat - (0xE000 - 0xD800) < 0xD800 := by grind
+    have : (0xE000 - 0xD800) ≤ x.toNat := by grind
+    simp [Char.ofNatAux, *]; rfl
+
+/-- Encode `Option (Fin n)` into `Fin (n+1)`. -/
+def encodeOption : Option (Fin n) → Fin (n+1)
+  | none => 0
+  | some ⟨i, h⟩ => ⟨i+1, Nat.succ_lt_succ h⟩
+
+/-- Decode `Option (Fin n)` from `Fin (n+1)`. -/
+@[pp_nodot] def decodeOption : Fin (n+1) → Option (Fin n)
+  | 0 => none
+  | ⟨i+1, h⟩ => some ⟨i, Nat.lt_of_succ_lt_succ h⟩
+
+@[simp] theorem encodeOption_decodeOption (x : Fin (n+1)) : encodeOption (decodeOption x) = x := by
+  simp only [encodeOption, decodeOption]
+  split
+  · next hd =>
+    split at hd
+    · next he => cases he; rfl
+    · contradiction
+  · next hd =>
+    split at hd
+    · contradiction
+    · next he => cases hd; simp
+
+@[simp] theorem decodeOption_encodeOption (x : Option (Fin n)) :
+    decodeOption (encodeOption x) = x := by
+  simp only [encodeOption, decodeOption]
+  split
+  · next he =>
+    split at he
+    · rfl
+    · cases he
+  · next he =>
+    split at he
+    · cases he
+    · cases he; rfl
+
+/-- Encode `Fin n ⊕ Fin m` into `Fin (n + m)`. -/
+def encodeSum : Sum (Fin n) (Fin m) → Fin (n + m)
+  | .inl x => x.castLE (Nat.le_add_right _ _)
+  | .inr x => x.natAdd n
+
+/-- Decode `Fin n ⊕ Fin m` from `Fin (n + m)`. -/
+@[pp_nodot] def decodeSum (x : Fin (n + m)) : Sum (Fin n) (Fin m) :=
+  if h : x < n then
+    .inl ⟨x, h⟩
+  else
+    .inr ⟨x - n, by grind⟩
+
+@[simp] theorem encodeSum_decodeSum (x : Fin (n + m)) : encodeSum (decodeSum x) = x := by
+  simp only [encodeSum, decodeSum]
+  split
+  · next hd =>
+    split at hd
+    · cases hd; rfl
+    · contradiction
+  · next hd =>
+    split at hd
+    · contradiction
+    · next he =>
+      cases x; cases hd
+      simp at he ⊢; grind
+
+@[simp] theorem decodeSum_encodeSum (x : Sum (Fin n) (Fin m)) : decodeSum (encodeSum x) = x := by
+  simp only [encodeSum, decodeSum]
+  split
+  · split
+    · simp
+    · next h => simp only [coe_castLE] at h; grind
+  · split
+    · next h => simp only [coe_natAdd] at h; grind
+    · next x _ => cases x; simp only [natAdd_mk, Sum.inr.injEq, mk.injEq]; grind
+
+/-- Encode `Fin m × Fin n` into `Fin (m * n)`. -/
+def encodeProd : Fin m × Fin n → Fin (m * n)
+  | (⟨i, hi⟩, ⟨j, hj⟩) => Fin.mk (n * i + j) <| calc
+    _ < n * i + n := Nat.add_lt_add_left hj ..
+    _ = n * (i + 1) := Nat.mul_succ ..
+    _ ≤ n * m := Nat.mul_le_mul_left n (Nat.succ_le_of_lt hi)
+    _ = m * n := Nat.mul_comm ..
+
+/-- Decode `Fin m × Fin n` from `Fin (m * n)`. -/
+@[pp_nodot] def decodeProd (z : Fin (m * n)) : Fin m × Fin n :=
```

Comment:
This duplicates `finProdFinEquiv` in Mathlib. That is fine, but I think at the very least one should consider porting over `divNat` and `modNat` from Mathlib for this purpose. I worry this is too "raw" a definition.

Reply:
This is the idea. These are low-level functions intended to replace the internal parts of `FinEquiv` in Mathlib. It's okay for these to be raw since they are not meant to be exposed to the general public. I do like `divNat` and `modNat` though.